### PR TITLE
Fix direct download of snapshots over http

### LINKED ISSUE #1936: fd_snapshot_loader can't follow snapshot redirects
Although fd_snapshot_http can follow redirects, fd_snapshot_loader can't.

fd_snapshot_loader attempts to extract the snapshot hash from the input string. 
It should try to parse it from the final redirect's URL instead of the first URL. 

```
NOTICE  05-23 15:54:35.499314 369660 6    replay:0 src/flamenco/snapshot/fd_snapshot.c(115): fd_snapshot_load(http://x.x.x.x:8899/incremental-snapshot.tar.bz2, verify-hash=false, check-hash=false, FD_SNAPSHOT_TYPE_INCREMENTAL)
ERR     05-23 15:54:35.499336 369660 6    replay:0 src/flamenco/snapshot/fd_snapshot.c(136): invalid snapshot hash
```

### PR DESCRIPTION
**Pull request by asiegel-jt**
Fixes #1936

### CODE REVIEW

File: src/flamenco/snapshot/fd_snapshot_http.c [hunk 1]
```diff
@@ -44,6 +44,7 @@ fd_snapshot_http_t *
 fd_snapshot_http_new( void *               mem,
                       uint                 dst_ipv4,
                       ushort               dst_port,
+                      ulong                base_slot,
```
Comment (ripatel-fd) [inline src/flamenco/snapshot/fd_snapshot_http.c:47]: I suggest moving this to 'set path'.

Comment (asiegel-jt) [reply src/flamenco/snapshot/fd_snapshot_http.c:47]: set_path happens deep in the api stack where base_slot isn't available

**Comment by ripatel-fd** [comment]
Marked as draft for now. Let's mark it ready to merge when it passes manual testing and is squashed.

File: src/flamenco/snapshot/fd_snapshot_http.h [hunk 2]
```diff
@@ -60,16 +62,36 @@ struct fd_snapshot_http {
...
+  /* Total downloaded so far */
+
+  ulong dl_total;
 };
 
 typedef struct fd_snapshot_http fd_snapshot_http_t;
 
 FD_PROTOTYPES_BEGIN
 
+/* TODO support case without a redirect (need to parse file name ahead of time) */
```
Comment (ripatel-fd) [inline src/flamenco/snapshot/fd_snapshot_http.h:88]: @asiegel-jt You fixed this right?

Comment (asiegel-jt) [reply src/flamenco/snapshot/fd_snapshot_http.h]: just did

File: src/flamenco/snapshot/fd_snapshot_loader.c [hunk 3]
```diff
@@ -131,19 +137,23 @@ fd_snapshot_loader_init( fd_snapshot_loader_t *    d,
       return NULL;
     }
 
+    if( FD_UNLIKELY( !fd_snapshot_name_from_cstr( &d->name, src->file.path, base_slot ) ) ) {
+      return NULL;
+    }
+
     if( FD_UNLIKELY( !fd_io_istream_file_new( d->vfile, d->snapshot_fd ) ) ) {
-      FD_LOG_WARNING(( "Failed to create fd_io_istream_file_t" ));
+      FD_LOG_WARNING(( "Failed to create fds_io_istream_file_t" ));
```
Comment (ripatel-fd) [inline src/flamenco/snapshot/fd_snapshot_loader.c:145]: typo

**Review by ripatel-fd — APPROVED** [review]

**Review by lidatong — APPROVED** [review]

**Pull request merged**